### PR TITLE
Improve reference resolution for deferred-annotations-within-classes

### LINKED ISSUE #1401: F401 (unused import) false positive
Hello, I have an import that is used in typing and is getting removed by ruff (flake8 detects it as unused as well) here is the example:

```python
from typing import Optional
import datetime # <--- gets removed by ruff
import pydantic

class SomeClass(pydantic.BaseModel):
    datetime: Optional[datetime.datetime]
```
for now, i'm adding the comment `# noqa: F401` to keep the import.

ruff version 0.0.195

Thank you for this project 🤩

### LINKED ISSUE #2248: F401+F811 false positive
#### Minimal reproducer
```python
from foo import Foo
from typing import TypeAlias

class Bar:
    Foo: TypeAlias = Foo
    def bar(self) -> Foo: ...
```

#### Output
```
$ ruff stub.pyi 
stub.pyi:1:17: F401 `foo.Foo` imported but unused
stub.pyi:5:5: F811 Redefinition of unused `Foo` from line 1
Found 2 error(s).
1 potentially fixable with the --fix option.
```

Ruff version: `0.0.227`

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

This PR modifies our symbol-resolution logic to better align with Pyright and other type checkers when handling deferred annotations in nested scopes.

By way of example, one of our oldest open issues points to Ruff raising a false-positive unused import here:

```py
from __future__ import annotations
from typing import Optional
import datetime # <--- gets removed by ruff
import pydantic

class SomeClass(pydantic.BaseModel):
    datetime: Optional[datetime.datetime]
```

The problem is that `Optional[datetime.datetime]` is evaluated last, since `from __future__ import annotations` is present. When we go to evaluate it, we see that `datetime` is already assigned in the scope, and resolve to _that_.

I did some research, and it looks like Pyright (and Mypy too) defer to the module scope when resolving such annotations, and fallback to resolving locally. For example, this works fine too:

```py
from typing import Optional, TypeAlias
import datetime # <--- gets removed by ruff
import pydantic

class SomeClass(pydantic.BaseModel):
    datetime: Optional[datetime.datetime]

    # X is defined locally, not in the module scope.
    X: TypeAlias = int
    y: X
```

This PR modifies our resolution logic to match Pyright's behavior, which in turn closes two long-standing issues. Note that this is a deviation from Pyflakes, but I think it's good to be more in-line with the type checkers.

Relevant Pyright issue: https://github.com/microsoft/pyright/issues/2644.

Relevant Pyright commit: https://github.com/microsoft/pyright/commit/27bf8c1f2.

Closes #1401.

Closes #2248.

### CODE REVIEW

File: crates/ruff_python_semantic/src/context.rs [hunk 1]
```diff
@@ -116,6 +116,17 @@ impl<'a> Context<'a> {
 
     /// Resolve a reference to the given symbol.
     pub fn resolve_reference(&mut self, symbol: &str, range: TextRange) -> ResolvedReference {
+        // PEP 563 indicates that if a forward reference can be resolved in the module scope, we
+        // should prefer it over local resolutions.
+        if self.in_deferred_type_definition() {
+            let scope_id = ScopeId::global();
+            if let Some(binding_id) = self.scopes[scope_id].get(symbol) {
```
Comment (MichaReiser) [inline crates/ruff_python_semantic/src/context.rs:123]:
```suggestion
            if let Some(binding_id) = self.scopes.global().get(symbol) {
```

**Review by MichaReiser — APPROVED** [review]

**Review by konstin — APPROVED** [review]

**Comment by sshishov** [comment]
Hello @charliermarsh ,

I do not know if this issue caused the new false-positive error in such code:

```python
from __future__ import annotations
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .. import some_module


def some_function(param: str) -> some_module.SupportedAdapters:
    from .. import some_module  # To fight with circular import, like this is util function  <-- this import is reported unused F401

    return cast(
            some_module.SupportedAdapters,
            import_string(f"some_path_here.adapters{param}")(value="init_value"),
        )
```

Or maybe the adapter pattern should use something else?

If everything is valid on `ruff` side, you can just add the comment here and ignore the issue. I do not want to create new unconfirmed issue, actually...

**Comment by charliermarsh** [comment]
Hmm, running `ruff check foo.py --select F --isolated` on that file does not yield an unused import error for me. Are you able to produce a reproducible file + command?

**Comment by sshishov** [comment]
Thanks @charliermarsh , this is the small example for reproduction (please note that this `TYPE_CHECKING` is required, you can try to delete it and you will get the circular import error).

NOTE: all these files are located inside the same package

**adapter.py**
```python
from __future__ import annotations

from utils import get_name


class Adapter:
    def __repr__(self) -> str:
        return f"{get_name()}()"
```

**main.py**
```python
from utils import get_adapter


if __name__ == "__main__":
    adapter = get_adapter()
    print(repr(adapter))
```

**utils.py**
```python
from __future__ import annotations

from typing import cast, TYPE_CHECKING

from django.utils.module_loading import import_string

if TYPE_CHECKING:
    from adapter import Adapter


def get_adapter() -> Adapter:
    from adapter import Adapter  # this import is marked unused, but deleting it will cause NameError

    return cast(Adapter, import_string("adapter.Adapter")())


def get_name() -> str:
    return "MyAdapter"
```

Run it out with/without problematic string:
```shell
╰─ python main.py
MyAdapter()

╰─ python main.py
Traceback (most recent call last):
  File "/home/user/ruff_example/main.py", line 4, in <module>
    adapter = get_adapter()
              ^^^^^^^^^^^^^
  File "/home/user/ruff_example/utils.py", line 14, in get_adapter
    return cast(Adapter, import_string("adapter.Adapter")())
                ^^^^^^^
NameError: name 'Adapter' is not defined
```

**Pull request merged**